ddtrace/contrib/starlette/patch.py

```diff
@@ -65,3 +53,42 @@ def unpatch():
     setattr(starlette, "_datadog_patch", False)
 
     _u(starlette.applications.Starlette, "__init__")
+
+    # We need to check that Fastapi instrumentation hasn't already unpatched these
+    if isinstance(starlette.routing.Route.handle, ObjectProxy):
+        _u(starlette.routing.Route, "handle")
+
+    if isinstance(starlette.routing.Mount.handle, ObjectProxy):
+        _u(starlette.routing.Mount, "handle")
+
+
+def traced_handler(wrapped, instance, args, kwargs):
+    # Since handle can be called multiple times for one request, we take the path of each instance
+    # Then combine them at the end to get the correct resource name
+    scope = get_argument_value(args, kwargs, 0, "scope")
+
+    if "__dd_paths__" in scope:
+        scope["__dd_paths__"].append(instance.path)
+
+    else:
+        scope["__dd_paths__"] = [instance.path]
+
+    if "datadog" not in scope:
+        return wrapped(*args, **kwargs)
+
+    # The scope["datadog"] object gets replaced with each new call to the handler
+    # Therefore we need to check if the current request span is the first one, and if it is, place it in __dd__
+    # So that it's not replaced and can be given the correct resource name
+    if "__dd__" not in scope:
+        scope["__dd__"] = {"__dd_first_request_span__": scope["datadog"].get("request_span")}
```

Comment:
I think we should probably fix this by not always overwriting the request span in the ASGI scope in the `TraceMiddleware` instead.

This is passable for now so I'll defer to other's opinions.

Reply:
yeah, what I am worried about is we now have 3 different keys we occupy in `scope`

`datadog`, `__dd__`, and `__dd_paths__`.

We should probably be tossing everything under `datadog`.

```python

if "datadog" not in scope:
    # Not actually sure how we should manage the actual root span vs the current handler request span
    scope["datadog"] = {"request_span": span, "root_request_span": span}
else:
    # Although, is this a problem once we unroll back to the root handler?
    scope["datadog"]["request_span"] = span


# In starlette
if "resource_paths" not in scope["datadog"]:
    scope["datadog"]["resource_paths"] = [path]
else:
    scope["datadog"]["resource_paths"].append(path)
```

Reply:
The problem with putting everything under `scope["datadog"]` is that it gets reset in the asgi middleware with each new handle method that's called. So if there's a subapp, it won't work. https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/contrib/asgi/middleware.py#L122

Reply:
> I'm not sure what behavior should be implemented so that we don't overwrite each time

that is the code sample/proposal in my last comment. in the asgi middleware, we should check for the presence of `scope["datadog"]` before we overwrite it, and instead update it accordingly for each sub-app being called.

I am not 100% sure the exact things that should be overwritten vs updated, or how we should design it, but we should be able to.

for example, do we overwrite the `scope["datadog"]["request_span"]` or do we treat it as a stack? ("stack" here meaning, we push the current request span into a list, and at the end of the request, we pop it off the list so when the sub-app is done being called, we can get the original parent request span back into the scope).


I might be overthinking this so would be worth getting your + others thoughts before we try to implement anything.

Reply:
Never mind, thought about it and implemented the change in the middleware. Let me know if it should be different.